### PR TITLE
Allow void entry point when calling BuildRunner.Execute.

### PR DESCRIPTION
**Pull request by adampoit**
There are two primary motivations for this change:

1. Enable unit testing Faithlife.Build in this repository. It is likely that we'd want to have test coverage for code in this repository, but disallowing void return entry points makes this challenging. By allowing void entry points we can write rich unit tests that validate behavior and prevent regressions.
2. Enable unit testing projects that depend on Faithlife.Build. It is not uncommon for us to ship common targets in a NuGet package that can be consumed by various projects. It would be very nice for us to be able to test those targets in a similar fashion. This is why I have chosen to publicly expose `BuildRunnerSettings`, instead of making it internal only.

I'm definitely open to feedback on whether this is the correct approach and API design.

### CODE REVIEW

File: src/Faithlife.Build/BuildRunner.cs [hunk 1]
```diff
@@ -16,22 +16,26 @@ public static class BuildRunner
 	/// </summary>
 	/// <param name="args">The command-line arguments from <c>Main</c>.</param>
 	/// <param name="initialize">Called to initialize the build.</param>
+	/// <param name="settings">The build runner settings.</param>
 	/// <returns>The exit code for the build.</returns>
-	public static int Execute(string[] args, Action<BuildApp> initialize) =>
-		ExecuteAsync(args, initialize).GetAwaiter().GetResult();
+	public static int Execute(string[] args, Action<BuildApp> initialize, BuildRunnerSettings? settings = null) =>
```
Comment (adampoit) [inline src/Faithlife.Build/BuildRunner.cs]: This is technically a binary breaking change. Would we prefer to add method overloads, or perhaps a method with a different name to allow void entry point execution?

Comment (ddunkin) [reply src/Faithlife.Build/BuildRunner.cs]: If the only use case here is unit testing, consider an internal overload.

File: src/Faithlife.Build/BuildRunner.cs [hunk 2]
```diff
@@ -16,22 +16,26 @@ public static class BuildRunner
...
-	public static async Task<int> ExecuteAsync(string[] args, Action<BuildApp> initialize)
+	public static async Task<int> ExecuteAsync(string[] args, Action<BuildApp> initialize, BuildRunnerSettings? settings = null)
 	{
+		settings ??= new BuildRunnerSettings();
+
 		ArgumentNullException.ThrowIfNull(args);
 		ArgumentNullException.ThrowIfNull(initialize);
 
-		if (Assembly.GetEntryAssembly()?.EntryPoint?.ReturnType == typeof(void))
+		if (!settings.AllowVoidEntrypoint && Assembly.GetEntryAssembly()?.EntryPoint?.ReturnType == typeof(void))
```
Comment (ejball) [inline src/Faithlife.Build/BuildRunner.cs]: Maybe we should just delete this surprising behavior. I forgot I added it. Long term, an analyzer would be a better solution to the problem of forgetting to return the exit code.

File: src/Faithlife.Build/BuildRunner.cs (hunk 1 above)
Comment (ejball) [reply src/Faithlife.Build/BuildRunner.cs]: We should avoid the binary breaking change, yes.

I don't mind `BuildRunnerSettings` if we want to maintain both behaviors.

File: tests/Faithlife.Build.Tests/BuildRunnerTests.cs [hunk 3]
```diff
@@ -15,4 +15,157 @@ public void NullInitializeThrows()
 	{
 		Assert.Throws<ArgumentNullException>(() => BuildRunner.Execute([], null!));
 	}
+
+	[Test]
+	public void FailsOnMissingTarget()
+	{
+		using var error = new StringWriter();
+		Console.SetError(error);
```
Comment (ejball) [inline tests/Faithlife.Build.Tests/BuildRunnerTests.cs:23]: I don't love the non-parallelizable tests, but I guess we don't have a choice.

I'm surprised they aren't being run in parallel, but maybe that's not the default.

File: src/Faithlife.Build/BuildRunner.cs (hunk 2 above)
Comment (adampoit) [reply src/Faithlife.Build/BuildRunner.cs]: Deleting the behavior would certainly remove any API changes and allow more flexibility in other repositories. If we're comfortable with that, that would likely be my preferred direction.

File: tests/Faithlife.Build.Tests/BuildRunnerTests.cs (hunk 3 above)
Comment (adampoit) [reply tests/Faithlife.Build.Tests/BuildRunnerTests.cs:23]: Another option would be to pass stdout/stderr streams via `BuildRunnerSettings` and have it prefer those over `Console` if they're present.

Comment (adampoit) [reply tests/Faithlife.Build.Tests/BuildRunnerTests.cs:23]: According to this, there is no parallel execution in NUnit by default: https://docs.nunit.org/articles/nunit/writing-tests/attributes/parallelizable.html

Comment (ejball) [reply tests/Faithlife.Build.Tests/BuildRunnerTests.cs:23]: That could work for direct Faithlife.Build output, but not for the apps that targets run, which could be confusing.

File: src/Faithlife.Build/BuildRunner.cs (hunk 2 above)
Comment (ejball) [reply src/Faithlife.Build/BuildRunner.cs]: It's fine by me.

Comment (adampoit) [reply src/Faithlife.Build/BuildRunner.cs]: In that case, I'm going to drop the API changes from this proposal.

File: tests/Faithlife.Build.Tests/BuildRunnerTests.cs (hunk 3 above)
Comment (adampoit) [reply tests/Faithlife.Build.Tests/BuildRunnerTests.cs:23]: Seeing as parallel execution is not the default in NUnit, I am going to say the `Console` redirection is fine as is.

**Review by ejball — APPROVED** [review]

**Pull request merged**